### PR TITLE
Spill join

### PR DESCRIPTION
**Pull request by pnowojski**
It covers issue: https://github.com/prestodb/presto/issues/5897

Before reviewing I highly recommend reading description of "Implemented solution" section of this doc: https://docs.google.com/document/d/1FyT2-2ocOZ6dvzGK70yrUIhd9oucmfnreFWqd3SmfyU/edit?usp=sharing#heading=h.jxm688ny3n2

### CODE REVIEW

**Comment by pnowojski** [comment]
@dain @cberner could one of you go through implementation description in google doc and validate high level design?

**Comment by cberner** [comment]
Yes, but before we move on to spilled joins, I'd like to see the follow-ups for spilled aggregations addressed, so that we have a solid implementation of them to base further work on.

**Comment by pnowojski** [comment]
@cberner we started to work on those follow ups, but they are parallel/independent of spilling in join. Since this is quite large PR, I would like to avoid rebasing it and maintaining if you will reject whole approach. 

Description of this implementation takes only a little bit more then 1 page on google docs:
https://docs.google.com/document/d/1FyT2-2ocOZ6dvzGK70yrUIhd9oucmfnreFWqd3SmfyU/edit#heading=h.jxm688ny3n2

**Comment by cberner** [comment]
Overall it seems like the right direction. I made some comments on details that I think should be changed. You may want to get @dain input as well, as he's most familiar with the details of our join operators

File: presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSourceFactory.java [hunk 1]
```diff
@@ -101,52 +142,221 @@ public PartitionedLookupSourceFactory(List<Type> types, List<Type> outputTypes,
...
-                else {
-                    this.lookupSourceSupplier = partitionLookupSource;
-                }
+            spilledLookupSources.put(partitionIndex, lookupSourceSpiller);
 
-                // store lookup source supplier and futures into local variables so they can be used outside of the lock
-                lookupSourceSupplier = this.lookupSourceSupplier;
-                lookupSourceFutures = ImmutableList.copyOf(this.lookupSourceFutures);
-            }
+            result = internalSetLookupSource(partitionIndex, () -> new SpilledLookupSource(types.size()));
```
Comment (ArturGajowy) [inline presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSourceFactory.java]: What's the point of storing SpilledLookupSource-s in a `Supplier<LookupSource>[]`?  Wouldn't `LookupSource[]` be enough? Or can the `types` change in the meantime and we're working around that?

**Comment by dain** [comment]
Please, assign this back to me when it is ready for review.

**Pull request closed without merging**